### PR TITLE
Eliminate GPU to CPU sync in linalg.eig for CUDA tensors

### LINKED ISSUE #167105: [feature request][linalg][CUDA] `torch.linalg.eig` -- remove host-device sync.
### 🚀 The feature, motivation and pitch

[#166715](https://github.com/pytorch/pytorch/pull/166715) (thank you, @johannesz-codes !), modified `torch.linalg.eig`'s logic to dispatch to cuSolver by default. cuSolver, similar to LAPACK/MAGMA, produces complex eigenvalues which require some processing to comply with the complex dtypes of PyTorch. The logic is handled by the method `linalg_eig_make_complex_eigenvectors`, which is implemented on the CPU.
That, of course, causes host-device syncs. Would be great to have it eliminated.

cc @ptrblck @msaroufim @eqy @jerryzh168 @jianyuh @mruberry @walterddr @xwang233 @Lezcano

### PR DESCRIPTION
**Pull request by razaaliraza**
This PR adds a CUDA kernel for linalg_eig_make_complex_eigenvectors,
eliminating unnecessary GPU→CPU→GPU synchronization when computing
eigenvalues and eigenvectors of real-valued CUDA tensors with complex eigenvalues.

When torch.linalg.eig() is called on a real-valued CUDA tensor that produces
complex eigenvalues, the implementation previously transferred eigenvectors to
CPU for decoding, then transferred them back to GPU. This PR keeps all
operations on the GPU, improving performance by ~8-10%.

## PR Overview

This PR makes linalg_eig_make_complex_eigenvectors dispatchable and adds a
CUDA kernel implementation. The changes span 4 commits:

**Commit 1: Add dispatch infrastructure**
- Declares DECLARE_DISPATCH stub for linalg_eig_make_complex_eigenvectors
- File: aten/src/ATen/native/BatchLinearAlgebra.h

**Commit 2: Refactor for dispatch**
- Moves CPU implementation to dispatch system
- Removes GPU-to-CPU synchronization from call site
- Preserves existing CPU implementation exactly (rename only)
- Files: aten/src/ATen/native/BatchLinearAlgebra.cpp, BatchLinearAlgebraKernel.cpp

**Commit 3: Add CUDA kernel**
- Implements parallel CUDA kernel for eigenvector decoding
- Parallelizes across batch, eigenvectors, and elements
- Uses AT_DISPATCH_V2 for float32/float64 support
- File: aten/src/ATen/native/cuda/BatchLinearAlgebraEig.cu

**Commit 4: Add comprehensive test**
- Tests CUDA kernel with rotation matrices (known analytical eigenvalues)
- Validates fundamental identity: A@v = λv
- Covers complex eigenvalues, real eigenvalues, and batched operations
- File: test/test_linalg.py

Now, for CUDA tensors, the entire pipeline runs on GPU with no host synchronization.

## Performance Results

Benchmarked on NVIDIA H200 with CUDA 12.8 using block-diagonal rotation matrices
(which guarantee complex eigenvalues, triggering the eigenvector decoding path).

Timing method: Wall-clock with GPU synchronization, 20 iterations after 5 warmup runs.

Configuration: batch_size × matrix_rows × matrix_cols
(e.g., 1×100×100 = single 100×100 matrix, 10×100×100 = batch of 10 matrices of size 100×100)

| Configuration  | Before (ms) | After (ms) | Improvement | Speedup |
|----------------|-------------|------------|-------------|---------|
| 1×100×100      |      11.459 |     10.506 |    0.953 ms |   8.3%  |
| 1×500×500      |      58.677 |     52.918 |    5.759 ms |   9.8%  |
| 1×1000×1000    |     123.130 |    110.919 |   12.211 ms |   9.9%  |
| 10×100×100     |     112.885 |    103.173 |    9.712 ms |   8.6%  |
| 100×100×100    |    1109.828 |   1024.709 |   85.119 ms |   7.7%  |

**Average speedup: ~9% for the entire linalg.eig() operation**

Fixes #167105

cc @ptrblck @msaroufim @eqy @jerryzh168 @tinglvv @nWEIdia @jianyuh @nikitaved @mruberry @walterddr @xwang233 @Lezcano

### CODE REVIEW

File: aten/src/ATen/native/cuda/BatchLinearAlgebraEig.cu [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    c10::complex<scalar_t>* __restrict__ result,
+    const c10::complex<scalar_t>* __restrict__ eigenvalues,
+    const scalar_t* __restrict__ vectors,
+    const int64_t batch_size,
+    const int64_t n,
+    const int64_t matrix_stride) {
+
+  const int64_t idx = blockIdx.x * blockDim.x + threadIdx.x;
+  const int64_t total_elements = batch_size * n * n;
+
```
Comment (lezcano) [inline aten/src/ATen/native/cuda/BatchLinearAlgebraEig.cu:28]: the number of entries should be `b * n^2` so this looks alright to me?

**Review by nikitaved — COMMENTED** [review]
nit: TensorIterator will likely lead to a much more compact PR. But I will not block it. Thoughts, @lezcano ?

File: aten/src/ATen/native/cuda/BatchLinearAlgebraEig.cu [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+  if (idx >= total_elements) return;
+
+  const int64_t batch_idx = idx / (n * n);
+  const int64_t local_idx = idx % (n * n);
+  const int64_t col = local_idx / n;
+  const int64_t row = local_idx % n;
```
Comment (nikitaved) [inline aten/src/ATen/native/cuda/BatchLinearAlgebraEig.cu:34]: The vector matrix is row-major? EDIT: the answer is no - col-major :)

File: aten/src/ATen/native/cuda/BatchLinearAlgebraEig.cu [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+  const auto n = real_vectors.size(-1);
+  const auto matrix_stride = matrixStride(real_vectors);
+  const auto batch_size = batchCount(real_vectors);
+
+  if (batch_size == 0 || n == 0) return;
+
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(complex_vectors.mT().is_contiguous());
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(complex_values.is_contiguous());
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(real_vectors.mT().is_contiguous());
```
Comment (nikitaved) [inline aten/src/ATen/native/cuda/BatchLinearAlgebraEig.cu:71]: 👍

File: test/test_linalg.py [hunk 4]
```diff
@@ -2267,6 +2267,90 @@ def test_eig_check_magma(self, device, dtype):
...
+
+        # Verify output is complex type
+        self.assertTrue(vals_complex.dtype in [torch.complex64, torch.complex128])
+        self.assertTrue(vecs_complex.dtype in [torch.complex64, torch.complex128])
+
+        # Verify Av = λv for both eigenpairs
+        for i in range(2):
+            lhs = A_complex.to(vecs_complex.dtype) @ vecs_complex[:, i]
+            rhs = vals_complex[i] * vecs_complex[:, i]
+            self.assertEqual(lhs, rhs, atol=1e-5, rtol=1e-5)
```
Comment (nikitaved) [inline test/test_linalg.py]: Why not compare `A @ v` to `lambda.unsqueeze(-2) * v`? That is vectorized and should handle batch dims just fine.

File: aten/src/ATen/native/cuda/BatchLinearAlgebraEig.cu (hunk 2 above)
Comment (razaaliraza) [reply aten/src/ATen/native/cuda/BatchLinearAlgebraEig.cu:34]: Its column major. Thanks!

**Review by lezcano — APPROVED** [review]
I think that TensorIterator is not necessary here, as we know the strides of the inputs and outputs.

That being said, the CUDA kernel is probably leaving quite a bit of juice on the table and it'd be a nice exercise on CUDA optimisation fixing it in a follow-up if @razaaliraza   is interested. Things that pop up are global the integer division and mod at the beginning of the kernel (vs. using multidimensional kernel launch), using `launch_bounds`, potentially looping over `n / block_size` elements in a for-loop and also have one thread handle two columns in one go if `imag > 0`...

That being said, this is certainly a good start, and for the other optimisations one should pull out ncu :)

File: test/test_linalg.py (hunk 4 above)
Comment (razaaliraza) [reply test/test_linalg.py]: Hi thank you for the suggestion, I have updated the test and pushed the changes.

File: test/test_linalg.py [hunk 5]
```diff
@@ -2267,6 +2267,87 @@ def test_eig_check_magma(self, device, dtype):
...
+            ], dtype=dtype, device=device),
+        ])
+
+        vals_batch, vecs_batch = torch.linalg.eig(A_batch)
+
+        # Verify Av = λv for each matrix in batch (vectorized inner loop)
+        for b in range(3):
+            lhs = A_batch[b].to(vecs_batch.dtype) @ vecs_batch[b]
+            rhs = vals_batch[b].unsqueeze(-2) * vecs_batch[b]
+            self.assertEqual(lhs, rhs, atol=1e-5, rtol=1e-5)
```
Comment (nikitaved) [inline test/test_linalg.py]: No need to loop over `b`, `A_batch @ vecs_batch` and `vals_batch.unsqueeze(-2) * vecs_batch` will handle batches just fine due to broadcasting.

Comment (razaaliraza) [reply test/test_linalg.py]: Yes, thanks again. Fixed and pushed.

File: test/test_linalg.py [hunk 6]
```diff
@@ -2267,6 +2267,87 @@ def test_eig_check_magma(self, device, dtype):
...
+        expected_val = torch.tensor(
+            expected_eigenvalue, dtype=vals_complex.dtype, device=device
+        )
+        expected_val_conj = torch.tensor(
+            expected_eigenvalue.conjugate(), dtype=vals_complex.dtype, device=device
+        )
+        self.assertTrue(
+            torch.allclose(vals_complex[0], expected_val, atol=1e-5, rtol=1e-5) or
+            torch.allclose(vals_complex[0], expected_val_conj, atol=1e-5, rtol=1e-5)
+        )
```
Comment (johannesz-codes) [inline test/test_linalg.py:2302]: Would it make sense to check both eigenvalues here? Since they are known to be ${\lambda, \overline{\lambda}}$ or  ${\overline{\lambda}, \lambda}$, validating both might improve the robustness of the test.

Comment (razaaliraza) [reply test/test_linalg.py:2302]: Hi @johannesz-codes thank you for the feedback, I agree with the observation and I have pushed the changes.

**Comment by razaaliraza** [comment]
I am trying to get access to an mi300 machine to debug/isolate this issue. Will update here. Thanks!

**Comment by Aidyn-A** [comment]
@razaaliraza, I think you are right, these ROCm failures are unrelated as these were flagged as flaky.

@pytorchbot merge

**Comment by nikitaved** [comment]
> Hi @Aidyn-A, looks like the CI is failing in a couple of cases on issues unrelated to this PR. Please let me know how to proceed. Thanks!

Hey! Sorry for the delay -- some CIs were disabled until very recently. Green light from me.

**Comment by razaaliraza** [comment]
Yeah, I can imagine things must be crazy right now. Thank you all for your time reviewing this!
